File: dotnet/Gitologist.Tests/PullTests.cs
using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace Gitologist.Tests;

[TestClass]
public class PullTests
{
    private string _testDir = null!;

    [TestInitialize]
    public void SetupTestDir()
    {
        _testDir = Path.Combine(
            Path.GetTempPath(),
            $"gitologist-pull-test-{DateTime.UtcNow.Ticks}-{Guid.NewGuid():N}"
        );
        Directory.CreateDirectory(_testDir);
    }

    [TestCleanup]
    public void CleanupTestDir()
    {
        if (Directory.Exists(_testDir))
        {
            try
            {
                Directory.Delete(_testDir, true);
            }
            catch
            {
            }
        }
    }

    [TestMethod]
    public async Task ShouldPullFromDefaultRemoteAndBranch()
    {
        await Init.InitRepo(_testDir);
        await File.WriteAllTextAsync(
            Path.Combine(_testDir, "test.txt"),
            "content"
        );
        await Add.AddFiles(_testDir, new[] { "test.txt" });
        await Commit.CreateCommit(_testDir, "Initial commit");

        await Push.PushToRemote(_testDir);

        await File.WriteAllTextAsync(
            Path.Combine(_testDir, "test.txt"),
            "modified"
        );
        await Add.AddFiles(_testDir, new[] { "test.txt" });
        await Commit.CreateCommit(_testDir, "Second commit");

        await Push.PushToRemote(_testDir);

        await Pull.PullFromRemote(_testDir);

        var content = await File.ReadAllTextAsync(
            Path.Combine(_testDir, "test.txt")
        );
        Assert.AreEqual("modified", content);
    }

    [TestMethod]
    public async Task ShouldPullFromSpecifiedRemote()
    {
        await Init.InitRepo(_testDir);
        await File.WriteAllTextAsync(
            Path.Combine(_testDir, "test.txt"),
            "content"
        );
        await Add.AddFiles(_testDir, new[] { "test.txt" });
        await Commit.CreateCommit(_testDir, "Initial commit");

        await Push.PushToRemote(_testDir, "upstream");

        await File.WriteAllTextAsync(
            Path.Combine(_testDir, "test.txt"),
            "modified"
        );
        await Add.AddFiles(_testDir, new[] { "test.txt" });
        await Commit.CreateCommit(_testDir, "Second commit");

        await Push.PushToRemote(_testDir, "upstream");

        await Pull.PullFromRemote(_testDir, "upstream");

        var content = await File.ReadAllTextAsync(
            Path.Combine(_testDir, "test.txt")
        );
        Assert.AreEqual("modified", content);
    }

    [TestMethod]
    public async Task ShouldPullFromSpecifiedBranch()
    {
        await Init.InitRepo(_testDir);
        var headPath = Path.Combine(_testDir, ".git", "HEAD");
        await File.WriteAllTextAsync(headPath, "ref: refs/heads/main");

        Directory.CreateDirectory(
            Path.Combine(_testDir, ".git", "refs", "heads")
        );
        await File.WriteAllTextAsync(
            Path.Combine(_testDir, ".git", "refs", "heads", "main"),
            "abc123"
        );

        await File.WriteAllTextAsync(
            Path.Combine(_testDir, "test.txt"),
            "content"
        );
        await Add.AddFiles(_testDir, new[] { "test.txt" });
        await Commit.CreateCommit(_testDir, "Initial commit");

        await Push.PushToRemote(_testDir, "origin", "main");

        await File.WriteAllTextAsync(
            Path.Combine(_testDir, "test.txt"),
            "modified"
        );
        await Add.AddFiles(_testDir, new[] { "test.txt" });
        await Commit.CreateCommit(_testDir, "Second commit");

        await Push.PushToRemote(_testDir, "origin", "main");

        await Pull.PullFromRemote(_testDir, "origin", "main");

        var content = await File.ReadAllTextAsync(
            Path.Combine(_testDir, "test.txt")
        );
        Assert.AreEqual("modified", content);
    }

    [TestMethod]
    public async Task ShouldThrowErrorIfNotAGitRepository()
    {
        var nonGitDir = Path.Combine(
            Path.GetTempPath(),
            $"not-a-repo-{DateTime.UtcNow.Ticks}"
        );
        Directory.CreateDirectory(nonGitDir);

        try
        {
            await Assert.ThrowsExceptionAsync<InvalidOperationException>(
                () => Pull.PullFromRemote(nonGitDir)
            );
        }
        finally
        {
            try
            {
                Directory.Delete(nonGitDir, true);
            }
            catch
            {
            }
        }
    }

    [TestMethod]
    public async Task ShouldThrowErrorIfRemoteBranchDoesNotExist()
    {
        await Init.InitRepo(_testDir);
        await File.WriteAllTextAsync(
            Path.Combine(_testDir, "test.txt"),
            "content"
        );
        await Add.AddFiles(_testDir, new[] { "test.txt" });
        await Commit.CreateCommit(_testDir, "Initial commit");

        await Assert.ThrowsExceptionAsync<InvalidOperationException>(
            () => Pull.PullFromRemote(_testDir)
        );
    }

    [TestMethod]
    public async Task ShouldUpdateLocalBranchReference()
    {
        await Init.InitRepo(_testDir);
        await File.WriteAllTextAsync(
            Path.Combine(_testDir, "test.txt"),
            "content"
        );
        await Add.AddFiles(_testDir, new[] { "test.txt" });
        await Commit.CreateCommit(_testDir, "Initial commit");

        await Push.PushToRemote(_testDir);

        await File.WriteAllTextAsync(
            Path.Combine(_testDir, "test.txt"),
            "modified"
        );
        await Add.AddFiles(_testDir, new[] { "test.txt" });
        var secondSha = await Commit.CreateCommit(_testDir, "Second commit");

        await Push.PushToRemote(_testDir);

        await Pull.PullFromRemote(_testDir);

        var localBranchPath = Path.Combine(
            _testDir,
            ".git",
            "refs",
            "heads",
            "main"
        );
        var localBranchContent = await File.ReadAllTextAsync(localBranchPath);

        Assert.AreEqual(secondSha, localBranchContent.Trim());
    }

    [TestMethod]
    public async Task ShouldHandleDirectories()
    {
        await Init.InitRepo(_testDir);
        Directory.CreateDirectory(Path.Combine(_testDir, "src"));
        await File.WriteAllTextAsync(
            Path.Combine(_testDir, "src", "index.ts"),
            "console.log('hello')"
        );
        await Add.AddFiles(_testDir, new[] { "src/index.ts" });
        await Commit.CreateCommit(_testDir, "Initial commit");

        await Push.PushToRemote(_testDir);

        await File.WriteAllTextAsync(
            Path.Combine(_testDir, "src", "index.ts"),
            "console.log('world')"
        );
        await Add.AddFiles(_testDir, new[] { "src/index.ts" });
        await Commit.CreateCommit(_testDir, "Second commit");

        await Push.PushToRemote(_testDir);

        await Pull.PullFromRemote(_testDir);

        var content = await File.ReadAllTextAsync(
            Path.Combine(_testDir, "src", "index.ts")
        );
        Assert.AreEqual("console.log('world')", content);
    }

    [TestMethod]
    public async Task ShouldHandleMultipleFiles()
    {
        await Init.InitRepo(_testDir);
        await File.WriteAllTextAsync(
            Path.Combine(_testDir, "file1.txt"),
            "content1"
        );
        await File.WriteAllTextAsync(
            Path.Combine(_testDir, "file2.txt"),
            "content2"
        );
        await File.WriteAllTextAsync(
            Path.Combine(_testDir, "file3.txt"),
            "content3"
        );
        await Add.AddFiles(
            _testDir,
            new[] { "file1.txt", "file2.txt", "file3.txt" }
        );
        await Commit.CreateCommit(_testDir, "Initial commit");

        await Push.PushToRemote(_testDir);

        await File.WriteAllTextAsync(
            Path.Combine(_testDir, "file1.txt"),
            "modified1"
        );
        await File.WriteAllTextAsync(
            Path.Combine(_testDir, "file2.txt"),
            "modified2"
        );
        await File.WriteAllTextAsync(
            Path.Combine(_testDir, "file3.txt"),
            "modified3"
        );
        await Add.AddFiles(
            _testDir,
            new[] { "file1.txt", "file2.txt", "file3.txt" }
        );
        await Commit.CreateCommit(_testDir, "Second commit");

        await Push.PushToRemote(_testDir);

        await Pull.PullFromRemote(_testDir);

        var content1 = await File.ReadAllTextAsync(
            Path.Combine(_testDir, "file1.txt")
        );
        var content2 = await File.ReadAllTextAsync(
            Path.Combine(_testDir, "file2.txt")
        );
        var content3 = await File.ReadAllTextAsync(
            Path.Combine(_testDir, "file3.txt")
        );

        Assert.AreEqual("modified1", content1);
        Assert.AreEqual("modified2", content2);
        Assert.AreEqual("modified3", content3);
    }

    [TestMethod]
    public async Task ShouldFastForwardWhenRemoteIsAhead()
    {
        await Init.InitRepo(_testDir);
        await File.WriteAllTextAsync(
            Path.Combine(_testDir, "test.txt"),
            "content"
        );
        await Add.AddFiles(_testDir, new[] { "test.txt" });
        var firstSha = await Commit.CreateCommit(_testDir, "Initial commit");

        await Push.PushToRemote(_testDir);

        await File.WriteAllTextAsync(
            Path.Combine(_testDir, "test.txt"),
            "modified"
        );
        await Add.AddFiles(_testDir, new[] { "test.txt" });
        var secondSha = await Commit.CreateCommit(_testDir, "Second commit");

        await Push.PushToRemote(_testDir);

        // Reset local branch back to first commit to simulate another clone
        var localBranchPath = Path.Combine(
            _testDir,
            ".git",
            "refs",
            "heads",
            "main"
        );
        await File.WriteAllTextAsync(localBranchPath, firstSha);
        await File.WriteAllTextAsync(
            Path.Combine(_testDir, "test.txt"),
            "content"
        );
        await Add.AddFiles(_testDir, new[] { "test.txt" });

        await Pull.PullFromRemote(_testDir);

        var content = await File.ReadAllTextAsync(
            Path.Combine(_testDir, "test.txt")
        );
        Assert.AreEqual("modified", content);

        var localBranchContent = await File.ReadAllTextAsync(localBranchPath);
        Assert.AreEqual(secondSha, localBranchContent.Trim());
    }

    [TestMethod]
    public async Task ShouldThrowErrorWhenLocalChangesWouldBeOverwritten()
    {
        await Init.InitRepo(_testDir);
        await File.WriteAllTextAsync(
            Path.Combine(_testDir, "test.txt"),
            "content"
        );
        await Add.AddFiles(_testDir, new[] { "test.txt" });
        var firstSha = await Commit.CreateCommit(_testDir, "Initial commit");

        await Push.PushToRemote(_testDir);

        await File.WriteAllTextAsync(
            Path.Combine(_testDir, "test.txt"),
            "modified"
        );
        await Add.AddFiles(_testDir, new[] { "test.txt" });
        await Commit.CreateCommit(_testDir, "Second commit");

        await Push.PushToRemote(_testDir);

        // Reset local branch back to first commit
        var localBranchPath = Path.Combine(
            _testDir,
            ".git",
            "refs",
            "heads",
            "main"
        );
        await File.WriteAllTextAsync(localBranchPath, firstSha);
        await File.WriteAllTextAsync(
            Path.Combine(_testDir, "test.txt"),
            "content"
        );
        await Add.AddFiles(_testDir, new[] { "test.txt" });

        // Make uncommitted local change
        await File.WriteAllTextAsync(
            Path.Combine(_testDir, "test.txt"),
            "local changes"
        );

        var exception = await Assert.ThrowsExceptionAsync<InvalidOperationException>(
            () => Pull.PullFromRemote(_testDir)
        );
        Assert.IsTrue(exception.Message.Contains("would be overwritten by merge"));

        // Verify local changes are preserved
        var content = await File.ReadAllTextAsync(
            Path.Combine(_testDir, "test.txt")
        );
        Assert.AreEqual("local changes", content);
    }

    [TestMethod]
    public async Task ShouldDeleteFilesRemovedOnRemote()
    {
        await Init.InitRepo(_testDir);
        await File.WriteAllTextAsync(
            Path.Combine(_testDir, "file1.txt"),
            "content1"
        );
        await File.WriteAllTextAsync(
            Path.Combine(_testDir, "file2.txt"),
            "content2"
        );
        await File.WriteAllTextAsync(
            Path.Combine(_testDir, "file3.txt"),
            "content3"
        );
        await Add.AddFiles(
            _testDir,
            new[] { "file1.txt", "file2.txt", "file3.txt" }
        );
        var firstSha = await Commit.CreateCommit(_testDir, "Initial commit");

        await Push.PushToRemote(_testDir);

        // Remove file2 on the remote
        File.Delete(Path.Combine(_testDir, "file2.txt"));
        await Add.AddAll(_testDir);
        await Commit.CreateCommit(_testDir, "Remove file2");

        await Push.PushToRemote(_testDir);

        // Reset local branch back to first commit to simulate another clone
        var localBranchPath = Path.Combine(
            _testDir,
            ".git",
            "refs",
            "heads",
            "main"
        );
        await File.WriteAllTextAsync(localBranchPath, firstSha);
        await File.WriteAllTextAsync(
            Path.Combine(_testDir, "file1.txt"),
            "content1"
        );
        await File.WriteAllTextAsync(
            Path.Combine(_testDir, "file2.txt"),
            "content2"
        );
        await File.WriteAllTextAsync(
            Path.Combine(_testDir, "file3.txt"),
            "content3"
        );
        await Add.AddFiles(
            _testDir,
            new[] { "file1.txt", "file2.txt", "file3.txt" }
        );

        await Pull.PullFromRemote(_testDir);

        // file2 should be deleted from the working tree
        Assert.IsFalse(File.Exists(Path.Combine(_testDir, "file2.txt")));

        // surviving files should be untouched
        var content1 = await File.ReadAllTextAsync(
            Path.Combine(_testDir, "file1.txt")
        );
        var content3 = await File.ReadAllTextAsync(
            Path.Combine(_testDir, "file3.txt")
        );
        Assert.AreEqual("content1", content1);
        Assert.AreEqual("content3", content3);
    }

    [TestMethod]
    public async Task ShouldPreserveLocallyModifiedFileDeletedOnRemote()
    {
        await Init.InitRepo(_testDir);
        await File.WriteAllTextAsync(
            Path.Combine(_testDir, "file1.txt"),
            "content1"
        );
        await File.WriteAllTextAsync(
            Path.Combine(_testDir, "file2.txt"),
            "content2"
        );
        await Add.AddFiles(_testDir, new[] { "file1.txt", "file2.txt" });
        var firstSha = await Commit.CreateCommit(_testDir, "Initial commit");

        await Push.PushToRemote(_testDir);

        // Remove file2 on the remote
        File.Delete(Path.Combine(_testDir, "file2.txt"));
        await Add.AddAll(_testDir);
        await Commit.CreateCommit(_testDir, "Remove file2");

        await Push.PushToRemote(_testDir);

        // Reset local branch back to first commit to simulate another clone
        var localBranchPath = Path.Combine(
            _testDir,
            ".git",
            "refs",
            "heads",
            "main"
        );
        await File.WriteAllTextAsync(localBranchPath, firstSha);
        await File.WriteAllTextAsync(
            Path.Combine(_testDir, "file1.txt"),
            "content1"
        );
        await File.WriteAllTextAsync(
            Path.Combine(_testDir, "file2.txt"),
            "content2"
        );
        await Add.AddFiles(_testDir, new[] { "file1.txt", "file2.txt" });

        // Make uncommitted local edits to file2, which was deleted on the remote
        await File.WriteAllTextAsync(
            Path.Combine(_testDir, "file2.txt"),
            "local changes"
        );

        await Pull.PullFromRemote(_testDir);

        // file2 should be preserved because it has uncommitted local edits
        var content2 = await File.ReadAllTextAsync(
            Path.Combine(_testDir, "file2.txt")
        );
        Assert.AreEqual("local changes", content2);

        // file1 should be untouched
        var content1 = await File.ReadAllTextAsync(
            Path.Combine(_testDir, "file1.txt")
        );
        Assert.AreEqual("content1", content1);
    }

    [TestMethod]
    public async Task ShouldNotOverwriteUnchangedFilesWithLocalModifications()
    {
        await Init.InitRepo(_testDir);
        await File.WriteAllTextAsync(
            Path.Combine(_testDir, "file1.txt"),
            "content1"
        );
        await File.WriteAllTextAsync(
            Path.Combine(_testDir, "file2.txt"),
            "content2"
        );
        await Add.AddFiles(_testDir, new[] { "file1.txt", "file2.txt" });
        var firstSha = await Commit.CreateCommit(_testDir, "Initial commit");

        await Push.PushToRemote(_testDir);

        // Commit a change only to file1
        await File.WriteAllTextAsync(
            Path.Combine(_testDir, "file1.txt"),
            "modified1"
        );
        await Add.AddFiles(_testDir, new[] { "file1.txt" });
        var secondSha = await Commit.CreateCommit(_testDir, "Second commit");

        await Push.PushToRemote(_testDir);

        // Reset local branch back to first commit
        var localBranchPath = Path.Combine(
            _testDir,
            ".git",
            "refs",
            "heads",
            "main"
        );
        await File.WriteAllTextAsync(localBranchPath, firstSha);
        await File.WriteAllTextAsync(
            Path.Combine(_testDir, "file1.txt"),
            "content1"
        );
        await File.WriteAllTextAsync(
            Path.Combine(_testDir, "file2.txt"),
            "content2"
        );
        await Add.AddFiles(_testDir, new[] { "file1.txt", "file2.txt" });

        // Make uncommitted change to file2 (which did not change in the pull)
        await File.WriteAllTextAsync(
            Path.Combine(_testDir, "file2.txt"),
            "local changes to file2"
        );

        await Pull.PullFromRemote(_testDir);

        // file1 should be updated
        var content1 = await File.ReadAllTextAsync(
            Path.Combine(_testDir, "file1.txt")
        );
        Assert.AreEqual("modified1", content1);

        // file2 local changes should be preserved since it wasn't changed in the pull
        var content2 = await File.ReadAllTextAsync(
            Path.Combine(_testDir, "file2.txt")
        );
        Assert.AreEqual("local changes to file2", content2);
    }
}

File: swift/Tests/GitologistTests/PullTests.swift
import Foundation
@testable import Gitologist
import Testing

struct PullTests {
	var testDir: URL {
		let tempDir = FileManager.default.temporaryDirectory
		let testName = "gitologist-test-\(Date().timeIntervalSince1970)-\(UUID().uuidString.prefix(8))"
		return tempDir.appendingPathComponent(testName)
	}

	let fileManager = FileManager.default

	init() {}

	@Test func shouldPullFromDefaultRemoteAndBranch() async throws {
		let testDirPath = testDir
		try fileManager.createDirectory(at: testDirPath, withIntermediateDirectories: true)
		try await initRepo(at: testDirPath.path)

		try "content".write(to: testDirPath.appendingPathComponent("test.txt"), atomically: true, encoding: .utf8)
		try await add(at: testDirPath.path, files: ["test.txt"])
		_ = try await commit(at: testDirPath.path, message: "Initial commit")

		try await push(at: testDirPath.path)

		try "modified".write(to: testDirPath.appendingPathComponent("test.txt"), atomically: true, encoding: .utf8)
		try await add(at: testDirPath.path, files: ["test.txt"])
		_ = try await commit(at: testDirPath.path, message: "Second commit")

		try await push(at: testDirPath.path)

		try await pull(at: testDirPath.path)

		let content = try String(contentsOf: testDirPath.appendingPathComponent("test.txt"), encoding: .utf8)
		#expect(content == "modified")

		try? fileManager.removeItem(at: testDirPath)
	}

	@Test func shouldPullFromSpecifiedRemote() async throws {
		let testDirPath = testDir
		try fileManager.createDirectory(at: testDirPath, withIntermediateDirectories: true)
		try await initRepo(at: testDirPath.path)

		try "content".write(to: testDirPath.appendingPathComponent("test.txt"), atomically: true, encoding: .utf8)
		try await add(at: testDirPath.path, files: ["test.txt"])
		_ = try await commit(at: testDirPath.path, message: "Initial commit")

		try await push(at: testDirPath.path, remote: "upstream")

		try "modified".write(to: testDirPath.appendingPathComponent("test.txt"), atomically: true, encoding: .utf8)
		try await add(at: testDirPath.path, files: ["test.txt"])
		_ = try await commit(at: testDirPath.path, message: "Second commit")

		try await push(at: testDirPath.path, remote: "upstream")

		try await pull(at: testDirPath.path, remote: "upstream")

		let content = try String(contentsOf: testDirPath.appendingPathComponent("test.txt"), encoding: .utf8)
		#expect(content == "modified")

		try? fileManager.removeItem(at: testDirPath)
	}

	@Test func shouldPullFromSpecifiedBranch() async throws {
		let testDirPath = testDir
		try fileManager.createDirectory(at: testDirPath, withIntermediateDirectories: true)
		try await initRepo(at: testDirPath.path)

		let headPath = testDirPath.appendingPathComponent(".git").appendingPathComponent("HEAD")
		try "ref: refs/heads/main\n".write(to: headPath, atomically: true, encoding: .utf8)

		try fileManager.createDirectory(
			at: testDirPath.appendingPathComponent(".git").appendingPathComponent("refs").appendingPathComponent("heads"),
			withIntermediateDirectories: true
		)
		try "abc123\n".write(
			to: testDirPath.appendingPathComponent(".git").appendingPathComponent("refs").appendingPathComponent("heads").appendingPathComponent("main"),
			atomically: true,
			encoding: .utf8
		)

		try "content".write(to: testDirPath.appendingPathComponent("test.txt"), atomically: true, encoding: .utf8)
		try await add(at: testDirPath.path, files: ["test.txt"])
		_ = try await commit(at: testDirPath.path, message: "Initial commit")

		try await push(at: testDirPath.path, branch: "main")

		try "modified".write(to: testDirPath.appendingPathComponent("test.txt"), atomically: true, encoding: .utf8)
		try await add(at: testDirPath.path, files: ["test.txt"])
		_ = try await commit(at: testDirPath.path, message: "Second commit")

		try await push(at: testDirPath.path, branch: "main")

		try await pull(at: testDirPath.path, branch: "main")

		let content = try String(contentsOf: testDirPath.appendingPathComponent("test.txt"), encoding: .utf8)
		#expect(content == "modified")

		try? fileManager.removeItem(at: testDirPath)
	}

	@Test func shouldThrowErrorIfNotAGitRepository() async throws {
		let tempDir = FileManager.default.temporaryDirectory
		let nonGitDir = tempDir.appendingPathComponent("not-a-repo-\(Date().timeIntervalSince1970)")
		try fileManager.createDirectory(at: nonGitDir, withIntermediateDirectories: true)

		await #expect(throws: PullError.self) {
			try await pull(at: nonGitDir.path)
		}

		try? fileManager.removeItem(at: nonGitDir)
	}

	@Test func shouldThrowErrorIfRemoteBranchDoesNotExist() async throws {
		let testDirPath = testDir
		try fileManager.createDirectory(at: testDirPath, withIntermediateDirectories: true)
		try await initRepo(at: testDirPath.path)

		try "content".write(to: testDirPath.appendingPathComponent("test.txt"), atomically: true, encoding: .utf8)
		try await add(at: testDirPath.path, files: ["test.txt"])
		_ = try await commit(at: testDirPath.path, message: "Initial commit")

		await #expect(throws: PullError.self) {
			try await pull(at: testDirPath.path)
		}

		try? fileManager.removeItem(at: testDirPath)
	}

	@Test func shouldUpdateLocalBranchReference() async throws {
		let testDirPath = testDir
		try fileManager.createDirectory(at: testDirPath, withIntermediateDirectories: true)
		try await initRepo(at: testDirPath.path)

		try "content".write(to: testDirPath.appendingPathComponent("test.txt"), atomically: true, encoding: .utf8)
		try await add(at: testDirPath.path, files: ["test.txt"])
		_ = try await commit(at: testDirPath.path, message: "Initial commit")

		try await push(at: testDirPath.path)

		try "modified".write(to: testDirPath.appendingPathComponent("test.txt"), atomically: true, encoding: .utf8)
		try await add(at: testDirPath.path, files: ["test.txt"])
		let secondSha = try await commit(at: testDirPath.path, message: "Second commit")

		try await push(at: testDirPath.path)

		try await pull(at: testDirPath.path)

		let localBranchPath = testDirPath.appendingPathComponent(".git").appendingPathComponent("refs").appendingPathComponent("heads").appendingPathComponent("main")
		let localBranchContent = try String(contentsOf: localBranchPath, encoding: .utf8)

		#expect(localBranchContent.trimmingCharacters(in: .whitespacesAndNewlines) == secondSha)

		try? fileManager.removeItem(at: testDirPath)
	}

	@Test func shouldHandleDirectories() async throws {
		let testDirPath = testDir
		try fileManager.createDirectory(at: testDirPath, withIntermediateDirectories: true)
		try await initRepo(at: testDirPath.path)

		try fileManager.createDirectory(at: testDirPath.appendingPathComponent("src"), withIntermediateDirectories: true)
		try "console.log('hello')".write(to: testDirPath.appendingPathComponent("src").appendingPathComponent("index.ts"), atomically: true, encoding: .utf8)
		try await add(at: testDirPath.path, files: ["src/index.ts"])
		_ = try await commit(at: testDirPath.path, message: "Initial commit")

		try await push(at: testDirPath.path)

		try "console.log('world')".write(to: testDirPath.appendingPathComponent("src").appendingPathComponent("index.ts"), atomically: true, encoding: .utf8)
		try await add(at: testDirPath.path, files: ["src/index.ts"])
		_ = try await commit(at: testDirPath.path, message: "Second commit")

		try await push(at: testDirPath.path)

		try await pull(at: testDirPath.path)

		let content = try String(contentsOf: testDirPath.appendingPathComponent("src").appendingPathComponent("index.ts"), encoding: .utf8)
		#expect(content == "console.log('world')")

		try? fileManager.removeItem(at: testDirPath)
	}

	@Test func shouldHandleMultipleFiles() async throws {
		let testDirPath = testDir
		try fileManager.createDirectory(at: testDirPath, withIntermediateDirectories: true)
		try await initRepo(at: testDirPath.path)

		try "content1".write(to: testDirPath.appendingPathComponent("file1.txt"), atomically: true, encoding: .utf8)
		try "content2".write(to: testDirPath.appendingPathComponent("file2.txt"), atomically: true, encoding: .utf8)
		try "content3".write(to: testDirPath.appendingPathComponent("file3.txt"), atomically: true, encoding: .utf8)
		try await add(at: testDirPath.path, files: ["file1.txt", "file2.txt", "file3.txt"])
		_ = try await commit(at: testDirPath.path, message: "Initial commit")

		try await push(at: testDirPath.path)

		try "modified1".write(to: testDirPath.appendingPathComponent("file1.txt"), atomically: true, encoding: .utf8)
		try "modified2".write(to: testDirPath.appendingPathComponent("file2.txt"), atomically: true, encoding: .utf8)
		try "modified3".write(to: testDirPath.appendingPathComponent("file3.txt"), atomically: true, encoding: .utf8)
		try await add(at: testDirPath.path, files: ["file1.txt", "file2.txt", "file3.txt"])
		_ = try await commit(at: testDirPath.path, message: "Second commit")

		try await push(at: testDirPath.path)

		try await pull(at: testDirPath.path)

		let content1 = try String(contentsOf: testDirPath.appendingPathComponent("file1.txt"), encoding: .utf8)
		let content2 = try String(contentsOf: testDirPath.appendingPathComponent("file2.txt"), encoding: .utf8)
		let content3 = try String(contentsOf: testDirPath.appendingPathComponent("file3.txt"), encoding: .utf8)

		#expect(content1 == "modified1")
		#expect(content2 == "modified2")
		#expect(content3 == "modified3")

		try? fileManager.removeItem(at: testDirPath)
	}
}
